src/array.ts

```diff
@@ -66,17 +66,24 @@ class ArrayType<T extends IAnyType> extends BaseType<Array<T["InputType"]> | und
     return value.every((child: any) => this.childrenType.is(child));
   }
 
-  instantiate(snapshot: this["InputType"] | undefined, context: InstantiateContext): this["InstanceType"] {
+  instantiate(snapshot: this["InputType"] | undefined, context: InstantiateContext, parent: IStateTreeNode | null): this["InstanceType"] {
```

Comment:
Now, the `instantiate` call for each type accepts the parent object calling it. This allows us to not have to indiscriminately try to set it on the result of any `.instantiate()` call, which often returns strings or other simple types, and instead have the type itself decide if it needs to care about the parent.